### Assignment 1 (3 scores):

- Use the Numpy library only to construct the Linear Regression model.
- Train and Evaluate that Linear Regression model on the [Forest Fires](https://archive.ics.uci.edu/static/public/162/forest+fires.zip) dataset.

In [1]:
import numpy as np
import pandas as pd

In [2]:
class Linear_Regression:
    def __init__(self):
        self.theta_ = None

    def rmse(self, y: np.ndarray, y_hat: np.ndarray) -> float:
        delta = y - y_hat
        return np.sqrt((delta**2).mean())

    def fit(self, X: np.ndarray, y: np.ndarray) -> None:
        cov_1 = np.matmul(X.T, X)
        inv_cov_1  = np.linalg.inv(cov_1)
        cov_2 = np.matmul(X.T, y)
        self.theta_ = np.matmul(inv_cov_1, cov_2)

    def predict(self, X: np.ndarray) -> np.ndarray:
        y_predict = X @ self.theta_.T
        return y_predict

In [3]:
data = pd.read_csv("forestfires.csv")
print(data.head(10))

   X  Y month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  area
0  7  5   mar  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.0
1  7  4   oct  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.0
2  7  4   oct  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.0
3  8  6   mar  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.0
4  8  6   mar  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.0
5  8  6   aug  sun  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.0
6  8  6   aug  mon  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.0
7  8  6   aug  mon  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.0
8  8  6   sep  tue  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.0
9  7  5   sep  sat  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.0


In [4]:
month = {
    "jan": 1,
    "feb": 2,
    "mar": 3,
    "apr": 4,
    "may": 5,
    "jun": 6,
    "jul": 7,
    "aug": 8,
    "sep": 9,
    "oct": 10,
    "nov": 11,
    "dec": 12,
}

data['month'] = data['month'].map(month) #convert month from str to num
print(data.head(10))

   X  Y  month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  area
0  7  5      3  fri  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.0
1  7  4     10  tue  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.0
2  7  4     10  sat  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.0
3  8  6      3  fri  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.0
4  8  6      3  sun  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.0
5  8  6      8  sun  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.0
6  8  6      8  mon  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.0
7  8  6      8  mon  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.0
8  8  6      9  tue  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.0
9  7  5      9  sat  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.0


In [5]:
day = {
    'mon': 2,
    'tue': 3,
    'wed': 4,
    'thu': 5,
    'fri': 6,
    'sat': 7,
    'sun': 8,
}

data['day'] = data['day'].map(day) #convert day from str to num
print(data.head(10))

   X  Y  month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  area
0  7  5      3    6  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.0
1  7  4     10    3  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.0
2  7  4     10    7  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.0
3  8  6      3    6  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.0
4  8  6      3    8  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.0
5  8  6      8    8  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.0
6  8  6      8    2  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.0
7  8  6      8    2  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.0
8  8  6      9    3  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.0
9  7  5      9    7  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.0


In [6]:
X = data.drop(columns = 'area')
Y = data['area'] #target

test_ratio = 0.2 #test size
np.random.seed(42)
indices = np.random.permutation(len(X))

X_new = X.iloc[indices]
Y_new = Y.iloc[indices]

#split X_new, Y_new to x_train, x_test, y_train, y_test
split_index = int(len(X_new) * (1 - test_ratio))
x_train = X_new.iloc[:split_index]
y_train = Y_new.iloc[:split_index]

x_test = X_new.iloc[split_index:]
y_test = Y_new.iloc[split_index:]
print(x_train.head(10))
print(y_train.head(10))


     X  Y  month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain
304  6  5      5    7  85.1   28.0  113.8   3.5  11.3  94   4.9   0.0
501  7  5      8    3  96.1  181.1  671.2  14.3  21.6  65   4.9   0.8
441  8  6      8    2  92.1  207.0  672.6   8.2  25.5  29   1.8   0.0
153  5  4      9    6  94.3   85.1  692.3  15.9  20.1  47   4.9   0.0
503  2  4      8    4  94.5  139.4  689.1  20.0  29.2  30   4.9   0.0
131  4  6      3    2  87.2   23.9   64.7   4.1  14.0  39   3.1   0.0
204  5  6      3    7  90.6   50.1  100.4   7.8  15.1  64   4.0   0.0
512  4  3      8    8  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0
325  1  4      9    7  92.2  102.3  751.5   8.4  24.2  27   3.1   0.0
247  2  4      8    4  92.2   91.6  503.6   9.6  20.7  70   2.2   0.0
304     0.00
501     0.00
441     1.23
153     1.46
503     1.95
131     0.00
204    13.99
512     6.44
325     0.00
247     0.75
Name: area, dtype: float64


In [7]:
model = Linear_Regression()
model.fit(x_train.values, y_train.values)
y_predict = model.predict(x_test.values)

#calculate RSME
rmse = model.rmse(y_test, y_predict)
print(rmse)

23.92838094910511


**REPORT FOR TASK 1:**
* **Model Implementation:** Successfully built the Linear Regression model from scratch using Numpy (implemented Normal Equation).
* **Data Preprocessing:** - Encoded string/categorical features (`month`, `day`) into numerical values.
  - Manually implemented data shuffling and an 80/20 train-test split without using `scikit-learn`.
* **Execution:** The model trained and predicted successfully on the test set.
* **Evaluation:** The initial RMSE on the test set is **23.93**. This value will serve as the baseline for our next optimization steps.

### Assignment 2 (3 scores):

- Standardize the data so that their mean is $0$ and their variance is $1$.
- Compare the results of Linear Regression model when being trained on the original and standardized data.

In [8]:
print(data.head(10))
data["X"] = np.

   X  Y  month  day  FFMC    DMC     DC   ISI  temp  RH  wind  rain  area
0  7  5      3    6  86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.0
1  7  4     10    3  90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.0
2  7  4     10    7  90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.0
3  8  6      3    6  91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.0
4  8  6      3    8  89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.0
5  8  6      8    8  92.3   85.3  488.0  14.7  22.2  29   5.4   0.0   0.0
6  8  6      8    2  92.3   88.9  495.6   8.5  24.1  27   3.1   0.0   0.0
7  8  6      8    2  91.5  145.4  608.2  10.7   8.0  86   2.2   0.0   0.0
8  8  6      9    3  91.0  129.5  692.6   7.0  13.1  63   5.4   0.0   0.0
9  7  5      9    7  92.5   88.0  698.6   7.1  22.8  40   4.0   0.0   0.0


In [ ]:
def Standard_z_score(X: np.ndarray):
    mean = np.mean(X)
    var = np.var(X)
    return

In [16]:
X_train = x_train.to_numpy()
print(type(x_train))
print(type(X_train))

<class 'pandas.DataFrame'>
<class 'numpy.ndarray'>


In [15]:
x_train["X"] = (x_train["X"] - np.mean(x_train["X"])) / np.var(x_train["X"])
x_train["Y"] = (x_train["Y"] - np.mean(x_train["Y"])) / np.var(x_train["Y"])
x_train["month"] = (x_train["month"] - np.mean(x_train["month"])) / np.var(x_train["month"])
x_train["day"] = (x_train["day"] - np.mean(x_train["day"])) / np.var(x_train[""])
x_train["FFMC"] = (x_train["FFMC"] - np.mean(x_train["FFMC"])) / np.var(x_train["FFMC"])
x_train["DMC"] = (x_train["DMC"] - np.mean(x_train["DMC"])) / np.var(x_train["DMC"])
x_train["DC"] = (x_train["DC"] - np.mean(x_train["DC"])) / np.var(x_train["DC"])
x_train["ISI"] = (x_train["ISI"] - np.mean(x_train["ISI"])) / np.var(x_train["ISI"])
x_train["temp"] = (x_train["temp"] - np.mean(x_train["temp"])) / np.var(x_train["temp"])
x_train["RH"] = (x_train["RH"] - np.mean(x_train["RH"])) / np.var(x_train["RH"])
x_train["wind"] = (x_train["wind"] - np.mean(x_train["wind"])) / np.var(x_train[""])
x_train["rain"] = (x_train["rain"] - np.mean(x_train["rain"])) / np.var(x_train["rain"])

KeyError: ''